hudi-utilities/src/test/java/org/apache/hudi/utilities/TestTimestampBasedKeyGenerator.java

```diff
@@ -0,0 +1,80 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.utilities;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hudi.DataSourceWriteOptions;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.util.SchemaTestUtil;
+import org.apache.hudi.common.util.TypedProperties;
+import org.apache.hudi.utilities.keygen.TimestampBasedKeyGenerator;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+
+public class TestTimestampBasedKeyGenerator {
+  private Schema schema = SchemaTestUtil.getTimestampEvolvedSchema();
+  private GenericRecord baseRecord = null;
+
+  public TestTimestampBasedKeyGenerator() throws IOException {
+  }
+
+  @Before
+  public void initialize() throws IOException {
+    baseRecord = SchemaTestUtil
+        .generateAvroRecordFromJson(schema, 1, "001", "f1");
+  }
+
+  private TypedProperties getBaseKeyConfig(String recordKeyFieldName, String partitionPathField, String hiveStylePartitioning) {
+    TypedProperties props = new TypedProperties();
+    props.setProperty(DataSourceWriteOptions.RECORDKEY_FIELD_OPT_KEY(), recordKeyFieldName);
+    props.setProperty(DataSourceWriteOptions.PARTITIONPATH_FIELD_OPT_KEY(), partitionPathField);
+    props.setProperty(DataSourceWriteOptions.HIVE_STYLE_PARTITIONING_OPT_KEY(), hiveStylePartitioning);
+    props.setProperty("hoodie.deltastreamer.keygen.timebased.timestamp.type", "EPOCHMILLISECONDS");
+    props.setProperty("hoodie.deltastreamer.keygen.timebased.output.dateformat", "yyyy-MM-dd hh");
+    props.setProperty("hoodie.deltastreamer.keygen.timebased.timezone", "GMT+8:00");
+    return props;
+  }
+
+  @Test
+  public void testTimestampBasedKeyGenerator() {
+    // if timezone is GMT+8:00
+    baseRecord.put("createTime", 1578283932000L);
+    TypedProperties props = getBaseKeyConfig("field1", "createTime", "false");
+    HoodieKey hk1 = new TimestampBasedKeyGenerator(props).getKey(baseRecord);
+    assertEquals(hk1.getPartitionPath(), "2020-01-06 12");
+
+    // if timezone is GMT
+    props.setProperty("hoodie.deltastreamer.keygen.timebased.timezone", "GMT");
+    HoodieKey hk2 = new TimestampBasedKeyGenerator(props).getKey(baseRecord);
+    assertEquals(hk2.getPartitionPath(), "2020-01-06 04");
+
+    // if timestamp is DATE_STRING, and timestamp type is DATE_STRING
+    baseRecord.put("createTime", "2020-01-06 12:12:12");
+    props.setProperty("hoodie.deltastreamer.keygen.timebased.timestamp.type", "DATE_STRING");
+    props.setProperty("hoodie.deltastreamer.keygen.timebased.input.dateformat", "yyyy-MM-dd hh:mm:ss");
+    props.setProperty("hoodie.deltastreamer.keygen.timebased.timezone", "GMT+8:00");
+    HoodieKey hk3 = new TimestampBasedKeyGenerator(props).getKey(baseRecord);
+    assertEquals(hk3.getPartitionPath(), "2020-01-06 12");
+  }
+}
```

Comment:
I think we would optimize the `TestTimestampBasedKeyGenerator` to the [gist](https://gist.github.com/leesf/b36654ba68c25a48d56958dabec07b83), WDYT?

Reply:
ok, What should I do?

Reply:
click the gist above.

Reply:
Is that right? This is the web address: [](https://gist.github.com/OpenOpened/2b50b86f21c36f3e07c78cb44d97ec9c)

Reply:
Is that right? This is the web address: [gist](https://gist.github.com/OpenOpened/2b50b86f21c36f3e07c78cb44d97ec9c)

Reply:
Then what do I need to do? addition more test unites?

Reply:
> Then what do I need to do? addition more test units?

You may copy the code from gist and cover the current `TestTimestampBasedKeyGenerator` class.

Reply:
already covered.